server/build.gradle

```diff
@@ -160,6 +160,9 @@ dependencies {
   //zstd
   api "com.github.luben:zstd-jni:${versions.zstd}"
 
+  // hashing
+  api "net.openhft:zero-allocation-hashing:${versions.zero_allocation_hashing}"
```

Comment:
@ketanv3 the improvements are non-negligible but I have concerns regarding this library:
 - it seems to be not actively maintained (the last commit is ~1 year ago)
 - is uses sun.misc.Unsafe, which is essentially a time bomb, it could block us from moving to recent JDKs any time

Have you considered any other implementation alternatives? (my apologies if I missed that)

Reply:
Thanks @reta! I do see your concerns.
* In all fairness to the library, given the static nature of hash functions, I don't expect the library to evolve much unless they add more hash functions or address existing bugs and security concerns.
* That's a fair point. I'll dig further to see what can be done.

> Have you considered any other implementation alternatives? (my apologies if I missed that)

The only other stable option is https://github.com/lz4/lz4-java#xxhash-java which is available in 3 implementations: JNI bindings, pure Java port and pure Java port that uses sun.misc.Unsafe. I remember benchmarking JNI and Unsafe version against OpenHFT and found it to be slower. I'll try the pure Java version and see how it stacks up.

The other idea I have is to port the required hash function from OpenHFT and rewrite it using VarHandles instead of Unsafe. It might be slightly slower (though JDK says it should be as performant), but should be a future-ready solution. What do you think?

Reply:
Thanks @ketanv3 

> The other idea I have is to port the required hash function from OpenHFT and rewrite it using VarHandles instead of Unsafe.

Could we make this change upstream? We could also try Project Panama (if it's helpful here), for example Apache Lucene already provides some vector processing based on it. In any case, non-Unsafe alternative would be indeed a future-ready solution, thank you.

Reply:
Hi @reta! The Vector API is still in early-access and the performance improvement may be platform-specific. Considering it requires a full re-write of the hash function (which is tricky to get right) and it benefits only a small number of users today (low adoption of the latest JDKs and experimental features), we can explore SIMD at a later point in time.

The OpenHFT library allows us to write a custom "accessor" to read primitives from a byte array. I wrote one that uses VarHandles instead of Unsafe, and the results are shared below.